src/ripple/consensus/Validations.h

```diff
@@ -362,148 +532,78 @@ class Validations
         return parms_;
     }
 
-    /** Return the journal
-     */
-    beast::Journal
-    journal() const
+    /** Return whether the local node can issue a validation for the given sequence
+        number
+
+        @param s The sequence number of the ledger the node wants to validate
+        @return Whether the validation satisfies the invariant, updating the
+                largest sequence number seen accordingly
+    */
+    bool
+    canValidateSeq(Seq const s)
     {
-        return j_;
+        ScopedLock lock{mutex_};
+        return localSeqEnforcer_(byLedger_.clock().now(), s, parms_);
     }
 
-    /** Result of adding a new validation
-     */
-    enum class AddOutcome {
-        /// This was a new validation and was added
-        current,
-        /// Already had this validation
-        repeat,
-        /// Not current or was older than current from this node
-        stale,
-        /// Had a validation with same sequence number
-        sameSeq,
-    };
-
     /** Add a new validation
 
         Attempt to add a new validation.
 
-        @param key The NodeKey to use for the validation
-        @param val The validation to store
-        @return The outcome of the attempt
-
-        @note The provided key may differ from the validation's
-              key() member since we might be storing by master key and the
-              validation might be signed by a temporary or rotating key.
+        @param key The master key associated with this validation
+        @param val The validationo to store
+        @return The outcome
 
+        @note The provided key may differ from the validations's  key()
+              member if the validator is using ephemeral signing keys.
     */
-    AddOutcome
+    ValStatus
     add(NodeKey const& key, Validation const& val)
     {
-        NetClock::time_point t = stalePolicy_.now();
-        if (!isCurrent(parms_, t, val.signTime(), val.seenTime()))
-            return AddOutcome::stale;
-
-        LedgerID const& id = val.ledgerID();
-
-        // This is only seated if a validation became stale
-        boost::optional<Validation> maybeStaleValidation;
-
-        AddOutcome result = AddOutcome::current;
+        if (!isCurrent(parms_, adaptor_.now(), val.signTime(), val.seenTime()))
+            return ValStatus::stale;
 
         {
             ScopedLock lock{mutex_};
 
-            auto const ret = byLedger_[id].emplace(key, val);
+            // Check that validation sequence is greater than any non-expired
+            // validations sequence from that validator
+            if (val.seq() != Seq{0})
+            {
+                auto const now = byLedger_.clock().now();
+                SeqEnforcer<Seq>& enforcer = seqEnforcers_[key];
+                if (!enforcer(now, val.seq(), parms_))
+                    return ValStatus::badSeq;
```

Comment:
So if there's multiple validations from a signer with the same seq but different ids, we discard all but the first one we see?

Reply:
That is correct. I suppose a better way to phrase the invariant as that a *signer* may only validate a ledger with larger sequence than any unexpired validation from that signer. So even if the signer rotates or revokes its signing keys and some new ledger with differing id is issued, we cannot use it if we already saw an earlier validation from that signer with the same seq.

Reply:
And "earlier validation" meaning one that we saw first (as opposed to `signTime` or anything else)